### ISSUE TITLE
Add an additional map pool

### ISSUE THREAD

**Issue by Bytekeeper** [Bytekeeper/basil#39]
In addition to the SSCAIT pool, add one that tries to follow the "official" pool with a delay.

Bots can signal (BASIL command) to use the "new" map pool if possible.
That means, if 2 bots are scheduled to play and both can play on the new map pool, a map from that pool will be chosen.
If one of the bots only supports SSCAITs pool, a map from that will be selected.
I expect this will affect ELO a bit, but not enough to warrant a second ELO.

**Comment by krasi0** [comment]
I find this a good compromise since it's opt-in :thumbsup:
This raises some additional questions: how often will the "official" map pool be changed and will previously opted-in bots be reverted to opt-out automatically after each update to the map pool?

**Comment by Bytekeeper** [comment]
1. The map pool change will not be often, as this is a manual process.
2. The choice is done by the author. It's opt-in. If the pool changes, I expected them to remove their "opt-in" or work already or get disabled after crashing for a while.

**Comment by chriscoxe** [comment]
TL;DR: suppose store a field for when each bot originally appeared and force them to support all map pools that were previously in use and the map pool currently in use, but also allow them to (voluntarily) write in their SSCAIT bot description "BASIL: MapPoolsUpTo=<a single number e.g. 3>" (the number must be >= the map pool number that was in use when they originally appeared) which means that they support all map pools up to and including 3.  When BotA plays against BotB, a map is selected from just the highest map pool number that both bots support (not any lower or higher map pool numbers than that).

Long version:

In principle, I don't think it is fair to judge a bot based on maps it plays that didn't exist/it didn't support when the bot was written, and it would devalue the rating system if a lot of existing bots are deleted or if they crash/hang/time out/perform worse because they are forced to play on unsupported maps, or if some bots can only play against a subset of other bots.  A bot should not boost the ratings of the other bots or cause erratic ratings changes or slow down or otherwise disrupt the server due to unsupported maps (e.g. frequently crashes/hangs/times out/breaks the host etc).

I propose calling the current map pool "map pool 1" and every time the latest map pool is expanded/reduced/changed, label the new set of maps with a new number (next "map pool 2").  So map pool 1 might be the current 14 maps, and map pool 2 could contain a subset of the maps from map pool 1 and some new maps (or a completely different set of maps), and similarly for map pool 3 etc.  I suggest allowing (but not requiring) SSCAIT bots' descriptions to contain "BASIL: MapPoolsUpTo=<a single integer> (e.g. "BASIL: MapPoolsUpTo=2" which means it must support all maps in map pools 1 and 2). Regardless, the bot MUST support all map pools up to and including the map pool number that was the latest map pool added to the server at the time they appeared. So, if they don't write this tag in their bot description, BASIL figures out which map pool was the latest at the time they originally appeared.  BASIL would reject and ignore "BASIL: MapPoolsUpTo=2" if map pool 3 was in use when the bot was originally appeared. BASIL would need a field for each bot to store when it originally appeared, and BASIL would also need to know the history when each map pool number went into use. If the bot is updated after that (e.g. to fix bugs in an old bot, like tscmoo did for Skynet), I don't think they should be forced to also add support for all the more recent map pools too.  Otherwise, this would be a disincentive to fix bugs/crashes/hangs/memory problems etc in old bots.

So suppose someone writes a new bot, after map pool 3 is added to the server.  When they first upload their bot, because map pool 3 was already added to the server, the rules for the server would say that new bots MUST support AT LEAST all the maps from the latest map pool (map pool 3 in this scenario) AND ALL EARLIER MAP POOLS (map pools 1 and 2 in this scenario) because it must be able to play against all old bots on a map that both bots support.

This allows old bots to be used and maintained, without too much onus on upgrading them/their terrain analysis libraries to support newer map pools, whilst allowing newer bots to play against each other on the highest numbered map pool that both boths support.  E.g. if BotA  has been configured to support map pools up to and including 2, and BotB has been configured to support map pools up to and including 3, then when BotA plays BotB, only the maps contained in map pool 2 would be sampled randomly to pick the map for that game.  I.E. Just a map from map pool 2, not from any maps in pool 1 that aren't also in map pool 2 (note that map pools may have some maps in common).

This way, for new bots, supporting the latest map pool as-at the time they originally appeared would not be opt-in - it would be a requirement.  I think it should be a requirement, not opt-in.  Of course, maintainers of old bots would still be able to switch their bot from "BASIL: MapPoolsUpTo=1" to "BASIL: MapPoolsUpTo=2" (I think this should be allowed, even though 3 is the latest map pool in this scenario) or "BASIL: MapPoolsUpTo=3", but I think this should be voluntary.  Also note that I don't think we should allow bots to only support map pools 1 and 3 (but not 2 as well) etc because it gives them a disincentive to use maps from map pool 2 if their bot supports all maps from maps pools 1, 2, 3 but plays better on maps from map pool 3 than map pool 2.  I think BASIL should allow the bot to switch from "Map pools 1, 2, 3" to "Map pool 1, 2" only if map pool 2 was the highest number of map pool that existed as-at when they originally appeared.

TBH, I think only a relatively small number of bots (compared to the full list of bots currently active on the server) will be updated to support the new map pool every time maps are added to the server.  Personally, when actively developing a bot and using the server to assess its performance/strengths/weaknesses, I would rather have a wide variety of opponents (even though most games are played on old maps) than a relatively small number of opponents on (mostly/only) the latest maps.

I think this proposal a good way to allow all old bots to just fix bugs etc AND allow 2 cutting edge bots to play against each other on just the latest map pool.

It is true that using my proposed rules, cutting-edge bots that originally appeared before the latest map pool number was added wouldn't have an incentive to switch to supporting the latest map pool unless they perform better on the latest map pool against bots that were added after that was added, but if they remained at the top of the ladder, I think the community would nudge them to add support for the more recent map pool(s), to join in the fun.

**Comment by chriscoxe** [comment]
If the map pool on BASIL and SSCAIT diverge, that would be a problem because authors might only be trying to support the SSCAIT maps, so if a new bot crashes on the BASIL-specific maps, a BASIL administrator could either manually edit the "when this bot appeared" field and set it to a date when map pool 1 was the latest map pool, or add an additional field that is used to specifically override the default logic (e.g. MapPoolsUpToOverride=1).

**Comment by chriscoxe** [comment]
Actually, I suppose as soon as the map pool on BASIL and SSCAIT diverge, it should be opt-in rather than a requirement.

**Comment by Bytekeeper** [comment]
Sorry, but this is too much hassle. There won't be more than 2 map pools. Since it's opt-in, no-one needs to take part.
I can however reset all bots back to using the SSCAIT map pool once an update was done. That way people will have to enable it manually once again or have their bot play on SSCAIT's map pool again.

There's no map pool divergence. Bots are expected to work on the SSCAIT pool, since they might play vs bots which do not support another pool.

**Comment by Bytekeeper** [comment]
This thing is live and working.
